File: Medium_Level_using_Python/Arrays_and_Strings/3. Longest Substring Without Repeating Characters.py
# Problem: 3. Longest Substring Without Repeating Characters
# Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
# Time Complexity: O(n)
# Space Complexity: O(k) - Size of character set- O(26) max
# Topics: Hash Table, String, Sliding Window
# Solution:

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chIndex = {}
        maxLen = 0
        start = 0

        for end in range(0, len(s)):
            if s[end] in chIndex and chIndex[s[end]] >= start:
                start = chIndex[s[end]] + 1
            chIndex[s[end]] = end
            maxLen = max(maxLen, end - start + 1)

        return maxLen

File: Medium_Level_using_Cpp/Array_and_String/3. Longest Substring Without Repeating Characters.cpp
// Problem: 3. Longest Substring Without Repeating Characters
// Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
// Time Complexity: O(n)
// Space Complexity: O(k) - Size of character set - max O(26)
// Topics: Hash Table, String, Sliding Window
// Solution: 

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndex;
        int maxLength = 0;
        int start = 0;

        for(int end = 0; end < s.length(); end++){
            if(charIndex.find(s[end]) != charIndex.end() && charIndex[s[end]] >= start){
                start = charIndex[s[end]] + 1;
            }
            charIndex[s[end]] = end;
            maxLength = max(maxLength, end - start + 1);
        }

        return maxLength;
    }
};